src/core/reference/include/openvino/reference/sparse_fill_empty_rows.hpp

```diff
@@ -0,0 +1,91 @@
+// Copyright (C) 2018-2025 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <algorithm>
+#include <unordered_set>
+#include <vector>
+
+#include "openvino/core/shape.hpp"
+
+namespace ov::reference {
+
+template <typename T, typename T_IDX>
+void sparse_fill_empty_rows(const T* values,
+                           const size_t values_size,
+                           const T_IDX* dense_shape,
+                           const T_IDX* indices,
+                           const size_t indices_size,
+                           const T default_value,
+                           T_IDX* output_indices,
+                           T* output_values,
+                           bool* empty_row_indicator) {
+    const T_IDX num_rows = dense_shape[0];
+    const size_t num_entries = values_size;
+
+    std::unordered_set<T_IDX> existing_rows;
+    for (size_t i = 0; i < num_entries; i++) {
+        // Only 2D inputs are supported, so every second idx is a row idx
+        existing_rows.insert(indices[i * 2]);
+    }
+
+    std::vector<T_IDX> empty_rows;
+    empty_rows.reserve(num_rows - existing_rows.size());
+    for (T_IDX i = 0; i < num_rows; i++) {
+        const bool is_empty = (existing_rows.find(i) == existing_rows.end());
+        empty_row_indicator[i] = is_empty;
+        if (is_empty) {
+            empty_rows.push_back(i);
```

Comment:
// Add entries for empty rows, first column
apply it here the `output_idx` should be `num_entries - 1` or `0`

Reply:
I changed the implementation quite a bit, I did not go the way you suggested, but instead the number of copies has been reduced, I removed the separate sorting step and now the output is populated in a single pass. If you'd prefer more optimizations let me know, I'll keep working on it. The changes I'm talking about were made in commit starting with `16b9579`.